src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs

```diff
@@ -0,0 +1,154 @@
+        /// </summary>
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <param name="httpClient">Instance of <see cref="HttpClient"/> to make the request with.</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static async Task<AuthenticationHeaderParser> ParseAuthenticationHeadersAsync(string resourceUri, CancellationToken cancellationToken, HttpClient httpClient)
+        {
+            if (httpClient is null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+            if (string.IsNullOrWhiteSpace(resourceUri))
+            {
+                throw new ArgumentNullException(nameof(resourceUri));
+            }
+
+            // call this endpoint and see what the header says and return that
+            HttpResponseMessage httpResponseMessage = await httpClient.GetAsync(resourceUri, cancellationToken).ConfigureAwait(false);
+
+            return ParseAuthenticationHeaders(httpResponseMessage.Headers);
+        }
+
+        /// <summary>
+        /// Creates a parsed set of parameters from the provided HttpResponseHeaders.
+        /// </summary>
+        /// <param name="httpResponseHeaders"></param>
+        /// <remarks>For known values, such as the nonce used for Proof-of-Possession, the parser will first chack for in the WWW-Authenticate headers
+        /// If it cannot find it, it will then check the Authentication-Info parameters for the value.</remarks>
+        /// <returns></returns>
+        public static AuthenticationHeaderParser ParseAuthenticationHeaders(HttpResponseHeaders httpResponseHeaders)
+        {
+            AuthenticationHeaderParser authenticationHeaderParser = new AuthenticationHeaderParser();
+            string serverNonce = null;
+
+            //Check for WWW-AuthenticateHeaders
+            if (httpResponseHeaders.WwwAuthenticate.Count != 0)
+            {
+                var WwwParameters = WwwAuthenticateParameters.CreateFromAuthenticateHeaders(httpResponseHeaders);
+
+                if (WwwParameters.Any(parameter => parameter.AuthScheme == "PoP"))
+                {
+                    serverNonce = WwwParameters.Where(parameter => parameter.AuthScheme == "PoP").Single().ServerNonce;
+                }
+
+                authenticationHeaderParser.ParsedWwwAuthenticateParameters = WwwParameters;
+            }
+            else
+            {
+                authenticationHeaderParser.ParsedWwwAuthenticateParameters = new List<WwwAuthenticateParameters>();
+            }
+
+            //If no WWW-AuthenticateHeaders exist, attempt to parse AuthenticationInfo headers instead
+            AuthenticationInfoParameters authenticationInfoParameters = AuthenticationInfoParameters.CreateFromHeaders(httpResponseHeaders);
```

Comment:
Should this be moved to the else block if this needs to be performed in case no www-authenticateHeaders are found?